orchagent/tunneldecaporch.cpp

```diff
@@ -0,0 +1,204 @@
+#include "tunneldecaporch.h"
+#include <string.h>
+#include "logger.h"
+
+extern sai_tunnel_api_t* sai_tunnel_api;
+extern sai_switch_api_t* sai_switch_api;
+
+TunnelDecapOrch::TunnelDecapOrch(DBConnector *db, string tableName) : Orch(db, tableName)
+{
+    SWSS_LOG_ENTER();
+}
+
+/**
+ * Function Description:
+ *    @brief reads from APP_DB and creates tunnel 
+ */
+void TunnelDecapOrch::doTask(Consumer& consumer)
+{
+    if (consumer.m_toSync.empty())
+        return;
+
+    auto it = consumer.m_toSync.begin();
+
+    while (it != consumer.m_toSync.end())
+    {
+        KeyOpFieldsValuesTuple t = it->second;
+
+        string key = kfvKey(t);
+        string op = kfvOp(t);
+
+        IpAddresses ip_addresses;
+        string tunnel_type;
+        string dscp_mode;
+        string ecn_mode;
+        string ttl_mode;
+        
+        if (op == SET_COMMAND)
+        {
+
+            for (auto i : kfvFieldsValues(t))
+            {
+                if (fvField(i) == "tunnel_type") 
+                {
+                    tunnel_type = fvValue(i);
+                }
+                if (fvField(i) == "dst_ip") 
+                {
+                    ip_addresses = IpAddresses(fvValue(i));
+                }
+                if (fvField(i) == "dscp_mode")
+                {
+                    dscp_mode = fvValue(i);
+                }
+                if (fvField(i) == "ecn_mode")
+                {
+                    ecn_mode = fvValue(i);
+                }
+                if (fvField(i) == "ttl_mode")
+                {
+                    ttl_mode = fvValue(i);
+                }
+            }
+            if (addDecapTunnel(tunnel_type, ip_addresses, dscp_mode, ecn_mode, ttl_mode)) 
+            {
+                SWSS_LOG_NOTICE("Tunnel(s) added to ASIC_DB.");
+            }
+            else
+            {
+                SWSS_LOG_ERROR("Failed to add tunnels to ASIC_DB.");
+            }
+        }
+
+        it = consumer.m_toSync.erase(it);
+    }
+}
+
+/**
+ * Function Description:
+ *    @brief adds a decap tunnel entry to ASIC_DB
+ *
+ * Arguments:
+ *    @param[in] type - type of tunnel
+ *    @param[in] dst_ip - destination ip address to decap
+ *    @param[in] dscp - dscp mode (uniform/pipe)
+ *    @param[in] ecn - ecn mode (copy_from_outer/standard)
+ *    @param[in] ttl - ttl mode (uniform/pipe)
+ *
+ * Return Values:
+ *    @return true on success and false if there's an error
+ */
+bool TunnelDecapOrch::addDecapTunnel(string type, IpAddresses dst_ip, string dscp, string ecn, string ttl)
+{
+
+    SWSS_LOG_ENTER();
+
+    int length_of_tunnel_attrs = 4;
+
+    // adding tunnel attributes to array and writing to ASIC_DB
+    sai_attribute_t tunnel_attrs[length_of_tunnel_attrs];
+
+    // tunnel type (only ipinip for now)
+    tunnel_attrs[0].id = SAI_TUNNEL_ATTR_TYPE;
+    tunnel_attrs[0].value.s32 = SAI_TUNNEL_IPINIP;
+
```

Comment:
looks like you need to set SAI_TUNNEL_ATTR_OVERLAY_INTERFACE for tunnel decap use. Before doing that, you need to craete a LOOPBACK type router interface, like below.

intf_attr.id = SAI_ROUTER_INTERFACE_ATTR_VIRTUAL_ROUTER_ID;
intf_attr.value.oid = default_virtual_router_id;
intf_attrs.push_back(intf_attr);
intf_attr.id = SAI_ROUTER_INTERFACE_ATTR_TYPE;
intf_attr.value.s32 = SAI_ROUTER_INTERFACE_TYPE_LOOPBACK;
intf_attrs.push_back(intf_attr);
create_router_interface(&underlay_if, intf_attrs.size(), intf_attrs.data());